zd3689082: Log charge id when making POST requests

oswaldquek opened the pull request:
We got quite a lot of logs in March like

`Connection timed out error for gateway url=...`
`Gateway returned unexpected status code: 503...`

These come from https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/gateway/GatewayClient.java.

It'd be useful to tie the chargeId to these logs so we can have another level
of confirmation what a charge's status should be from the calls that were made.

alexbishop1 commented:
Most of the time, we should output a representation of the gateway response (see https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseService.java#L91 and https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthService.java#L79 and a few similar lines) but it does require us to know what to look for in the response and I’m not sure this is consistently done for every gateway (I’m not sure Stripe responses get logged like that) and possibly not for Apple Pay and Google Pay. Logging the whole response seems fine but we have to be careful we don’t accidentally log personally-identifiable information.

rhowe-gds commented on src/main/java/uk/gov/pay/connector/paymentprocessor/resource/CardResource.java:
```diff
@@ -3,6 +3,7 @@
 import com.google.common.collect.ImmutableMap;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
```
`pay-java-commons` uses jboss's MDC. I don't know if mixing them is a good idea?
We should probably switch to using the slf4j one? I don't know what if any difference there is

oswaldquek replied:
Do you mean switching the one in `pay-java-commons`? I don't think there's a problem with this line as we do a similar thing at https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureProcess.java#L53

rhowe-gds commented:
If you add something to the MDC my understanding is that it will persist for the lifetime of the thread. This presumably means that following requests may inherit a `chargeId` of a previous request unless we guarantee that every request to connector sets the `chargeId` key in the MDC

We may be setting ourselves up for a world of pain here - using the power of Splunk we should be able to tie things together by request ID instead?

oswaldquek commented:
@rhowe-gds I think you're right about following requests inheriting a `chargeId`. Looking at https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureProcess.java#L94 it looks like the requestID is removed. This means I'll have to do a similar thing but it means writing `try...finally`s everywhere :(

oswaldquek commented:
@rhowe-gds I don't think we can use the requestID.... it's already being logged and I can't tell what external requests to PSPs relate to which charge ids.

oswaldquek commented:
@rhowe-gds maybe a better solution is to to `MDC.put("chargeId", chargeId)` in https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/gateway/GatewayClient.java#L56 instead? Then we can remove it in the finally block?

oswaldquek commented:
@alexbishop1 you're right. https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthService.java#L85 already logs the raw response from the PSP. i've removed the code i've added to log the response from the server.

rhowe-gds commented:
> @rhowe-gds maybe a better solution is to to `MDC.put("chargeId", chargeId)` in https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/gateway/GatewayClient.java#L56 instead? Then we can remove it in the finally block?

Seems fair. You only need to use the MDC if you want a shorthand way of passing the chargeId around though - if the logs you want the chargeId to be associated with are logged by that method we can just include it in the log message and the MDC doesn't need to know about it at all.

heathd commented:
> We may be setting ourselves up for a world of pain here - using the power of Splunk we should be able to tie things together by request ID instead?

I think that's a bit optimistic unfortunately. When you're trying to look at patterns of behaviour, tying things to context via the requestID is challenging. ChargeID is so ubiquitous that I think there's good reason to have it on every log line where it's appropriate.

In the ideal world that would mean:

- for HTTP request we should set it at the start of the HTTP request and clear it at the end
- for background threads (as is the case here) we should set it when the thread starts processing a charge and clear it when done

As Russell said there the MDC is a thread local so you do need to be careful to clear it, and I think using a try/finally block is the only reliable way to do that. 

For the case of this PR, hooking into a try/finally block in the gateway client seems reasonable

alexbishop1 commented:
@oswaldquek Well, it’s not quite the raw response (it’s the stringification of the Java object we deserialise the response into) but it’s usually pretty good (we’ve improved it a few times to extract more relevant information). For ePDQ payments, it gets the status and the error messages, which is the main thing.

oswaldquek commented:
@alexbishop1 oh, I was wrong.  https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthService.java#L85 DOESN'T log the raw response from the PSP. it's because our error handling code is ridiculously hard to read 💢

oswaldquek commented:
AOP sounds good for this type of situation. Sadly it doesn't work out of the box with dropwizard 😢 
I'll investigate doing this in a filter.

heathd commented on src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardExecutorService.java:
```diff
@@ -109,6 +109,7 @@ public ExecutorService getExecutor() {
                 return task.call();
             } finally {
                 xrayUtils.endSegment();
+                MDC.clear();
```
ok this may be a bit picky but I think:
- the try{ should enclose line 100 `setContextMap` 
- possibly we should store the MDC before setting and then restore it?

oswaldquek replied:
hmm does it matter? we have this logged on line 103: `logger.debug("Card operation task spent {} ms in queue", totalWaitTime);` do we not want to log the chargeId there too?

oswaldquek replied:
alternatively we could move the `try {` to line 100, i.e. before the `MDC.setContextMap(mdcContextMap);`

heathd replied:
```
    // accepts a supplier function and executed that in a separate Thread of its own.
    // returns a Pair of the execution status and the return type
    public <T> Pair<ExecutionStatus, T> execute(Supplier<T> callable) {
        Callable<T> task = callable::get;
        Map<String, String> mdcContextMap = Optional.ofNullable(MDC.getCopyOfContextMap()).orElse(Map.of());
        final long startTime = System.currentTimeMillis();

        Future<T> futureObject = executor.submit(() -> {
            try {
              MDC.setContextMap(mdcContextMap);
              xrayUtils.beginSegment();
              long totalWaitTime = System.currentTimeMillis() - startTime;
              logger.debug("Card operation task spent {} ms in queue", totalWaitTime);
              if (totalWaitTime > QUEUE_WAIT_WARN_THRESHOLD_MILLIS) {
                  logger.warn("CardExecutor Service delay - queue_wait_time={}", totalWaitTime);
              }
              metricRegistry.histogram("card-executor.delay").update(totalWaitTime);
              return task.call();
            } finally {
                xrayUtils.endSegment();
                MDC.clear();
            }
        });
```
